main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala

```diff
@@ -15,33 +15,41 @@
  */
 
 package ca.uwaterloo.flix.api.lsp.provider
-import ca.uwaterloo.flix.api.lsp.acceptors.{FileAcceptor, AllAcceptor}
+import ca.uwaterloo.flix.api.lsp.acceptors.{FileAcceptor}
 import ca.uwaterloo.flix.language.ast.shared.SymUse
-import ca.uwaterloo.flix.language.ast.SourceLocation
-import ca.uwaterloo.flix.language.ast.TypedAst.{Expr, Op, Root}
+import ca.uwaterloo.flix.language.ast.TypedAst.{Expr, Root}
 import ca.uwaterloo.flix.api.lsp.{Consumer, InlayHint, InlayHintKind, Position, Range}
 import ca.uwaterloo.flix.api.lsp.Visitor
 
-
 object InlayHintProvider {
-  private val EnableEffectHints: Boolean = true
+  private val EnableEffectHints: Boolean = false
 
   def getInlayHints(uri: String, range: Range)(implicit root: Root): List[InlayHint] = {
     var inlayHints: List[InlayHint] = Nil
     if(EnableEffectHints) {
-      var opSymUses: Set[(SymUse.OpSymUse, SourceLocation)] = getOpSymUses(uri)
-      opSymUses.foreach { case (opSymUse, loc) => mkHintFromOpSymUse(opSymUse, loc) :: inlayHints }
+      var effects: Set[(String, Int, Int)] = (getOpSymUses(uri) ++ getDefSymUses(uri)).filter {
+        case (eff, _, _) =>
+          eff != "Pure"
+      }
+      var maxColWidth: Int = effects.map { case (_, _, col) => col }.maxOption.getOrElse(0)
+      var lineToEffectsMap: Map[Int, Set[String]] = Map.empty
+      effects.foreach { case (eff, line, col) =>
+        lineToEffectsMap = lineToEffectsMap.updated(line, lineToEffectsMap.getOrElse(line, Set.empty) + eff)
+      }
+      lineToEffectsMap.foreach { case (line, effects) =>
+        inlayHints ::= mkHintFromEffects(line, effects, maxColWidth + 1)
+      }
```

Comment:
I would introduce a handful of functions for the above functionality. Also try to avoid imperative programming, but use functional programming.